### DISCUSSION TITLE
Clarify Roadmap/Versions and Upgrades

### DISCUSSION THREAD

**Discussion by RegisWeduLabs** [harvester/harvester#5826] [General]
Hi,

I'm opening this discussion as it's not a Feature/Bug, but rather a general matter. I hope this is the appropriate forum for it.

First and foremost, I want to express my gratitude for Harvester. I admire the level of skill and technology behind this project. It's truly impressive.

That said, I do have some concerns about the project's general direction. As with most users, this is a critical aspect of our operations, and I believe transparency and a clear roadmap/goals are crucial.

1. **Upgrade Process**: I raised an issue about 2 years ago (https://github.com/harvester/harvester/issues/3053) - and even though it was closed, I still feel like the current upgrade mechanism has many downsides, making upgrades a very stressful operation. There is an EPIC (https://github.com/harvester/harvester/issues/4926) about enhancing upgrades, but it doesn't really address previous concerns. During an upgrade, many components are affected, troubleshooting can be quite challenging, rollbacks are nearly impossible, and there is no way to suspend the upgrade if it goes wrong. The process is also very lengthy, even if all goes well.

2. **Release Notes/Upgrade Support Matrix**: When 1.3.0 was announced, it brought many long-awaited features, only to realize that upgrading from 1.2.1 to 1.3.0 is not possible. This was a bit of a shock, considering most users are likely using 1.2.1. The release notes stated that upgrading to 1.3.0 would be possible once 1.2.2 is released, but now the 1.2.2 release notes say it will support upgrading to 1.3.1 instead. This lack of clarity or contradiction between releases (1.3.0 and 1.2.2) is confusing. The upgrade support matrix should clearly state what is possible and what isn't for current users looking to upgrade.

3. **Roadmap Clarity**: The roadmap (https://github.com/harvester/harvester/wiki/Roadmap) shows that version 1.2.2 was released in May, suggesting an upgrade path of 1.2.1->1.2.2->1.3.1. However, there is also a 1.2.3 release listed in the milestones for July after 1.3.1 releases in June. If the upgrade path is 1.2.1->1.2.2->1.3.1, why is there a 1.2.3 release, and what does it mean for the upgrade path? Does this indicate a "fork" in the upgrade path, similar to Rancher, where there are parallel releases? This information is crucial for planning upgrades, especially since we need to schedule maintenance and inform users weeks in advance. It is also important to understand the potential security risks for Harvester users in delaying an upgrade to combine 2-3 releases in one go (next point).

4. **Security Concerns**: Not upgrading the OS/apps for months can be risky as new vulnerabilities are discovered. While some deployments are air-gapped, others are internet-facing. Recently discovered CVEs in Rancher, which Harvester relies on, cannot be addressed until the next full Harvester release. There was also no guidance on the potential impact on current Harvester deployments or suggestions on minimizing the risk.

Again, this really is meant to be constructive and open an exchange/discussion. I understand Harvester very well and am aware of the complexity of the above issues. I believe clearer communications and a transparent roadmap can address most of these concerns, while technical challenges will improve as Harvester matures.

Thank you for your consideration and all the good work put into this project.

**Comment by tserong** [comment]
Thank you for the feedback!

Regarding point 2, upgrading 1.2.1 to 1.3.0 wasn't possible due to the k8s requirement of upgrading one minor version at a time. Harvester 1.2.1 uses RKE2 v1.25.9+rke2r1 and Harvester 1.3.0 uses RKE2 v1.27.10+rke2r1. We needed to go to Harvester 1.2.2 first (which has RKE2 v1.26.15+rke2r1) in order to correctly work through v1.25.x->v1.26.x->v.1.27.x of k8s. Unfortunately, we later found a live migration issue in Harvester v1.2.2's version of kubevirt, which also had to be fixed in the v1.3.x series, hence the need to upgrade to v1.3.1, not v1.3.0. This has since been updated in the Upgrade section of the [v1.3.0](https://github.com/harvester/harvester/releases/tag/v1.3.0) and [v1.2.2](https://github.com/harvester/harvester/releases/tag/v1.2.2) release notes.

Point 3 has been added as a separate discussion item (https://github.com/harvester/harvester/discussions/6009) and @bk201 has replied there.

Points 1 and 4 are arguably related - they both come down to upgrade granularity, and as you know this is complicated :-) On the one hand, if you can upgrade individual components independently it can (e.g. as you suggest) make troubleshooting easier, as only one thing has changed, and could help to alleviate concerns about CVEs. On the other hand, we need to be able to test and verify to ensure that any individual update will not break the Harvester cluster - if we work with a known set of component versions, as we have in each Harvester release, and upgrade all the components at the same time, it makes the test matrix much smaller (and thus achievable to test!) than it would be if we had to allow any component to be upgraded independently. Of course, there may be a happy middle ground between the two extremes that we just still need to find, so IMO this remains a topic for further consideration.

**Reply by RegisWeduLabs** [reply]
Thank you @tserong - very appreciated.

I still think that:

1) Yes a middle ground would be very nice - especially for Rancher (in harvester) and Rancher that manages harvester -  and the OS itself - as those are the 2 components that are most prone to CVEs.  Depending on the criticality of the discovered vulnerability, it might be quite critical to upgrade promptly.  While I understand that Rancher documents and release the fixes, Harvester also uses Rancher extensively - and is also most likely affected.  Stripped down Rancher in harvester and the Linux Kernel should be patchable/upgradable separatly to mitigate most of the potential public-facing CVEs or even to fix a critical bug.

2) Transparency and communication is key - if there is a known critical/dangerous vulnerability discovered and posted for everyone to see, there should be some sort of communication to the community about 1 - acknowledging the risk and 2 - give instructions on what they need to do in order to mitigate/workaround while upgrades are being worked on.  This is critical in keeping HCI environments secured - since some of the services are internet facing.  

Of course, I do understand the complexity, but I don't think this is much of a choice if Harvester is to be adopted on a larger scale.

Thank you!
